hub/core/storage/s3/s3_mapper.py

```diff
@@ -0,0 +1,173 @@
+import posixpath
+import boto3
+import botocore
+from typing import Optional
+from collections.abc import MutableMapping
+from botocore.exceptions import ClientError
+from hub.util.exceptions import S3GetError, S3SetError, S3DeletionError, S3ListError
+
+
+class S3Mapper(MutableMapping):
+    """
+    An s3 mapper built using boto3. For internal use only (by class S3Provider)
+    """
+
+    def __init__(
+        self,
+        url: str,
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        aws_session_token: Optional[str] = None,
+        endpoint_url: Optional[str] = None,
+        aws_region: Optional[str] = None,
+        max_pool_connections: Optional[int] = 25,
+    ):
+        self.aws_region = aws_region
+        self.endpoint_url = endpoint_url
+
+        self.bucket = url.split("/")[0]
+        self.path = "/".join(url.split("/")[1:])
+
+        self.client_config = botocore.config.Config(
+            max_pool_connections=max_pool_connections,
+        )
+
+        self.client = boto3.client(
+            "s3",
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+            aws_session_token=aws_session_token,
+            config=self.client_config,
+            endpoint_url=self.endpoint_url,
+            region_name=self.aws_region,
+        )
+
+    def __setitem__(self, path, content):
+        """
+        Sets the object present at the path with the value
+
+        Args:
+            path (str): the path relative to the root of the mapper.
+            content (bytes): the value to be assigned at the path.
+
+        Returns:
+            None
+
+        Raises:
+            S3SetError: Any S3 error encountered while setting the value at the path.
+        """
+        try:
+            path = posixpath.join(self.path, path)
```

Comment:
What exactly does `posixpath.join` do? Is it just a fancy concat for two paths?

Reply:
>On Mac/Linux, os.path.join is an alias for posixpath.join, which always joins path segments with /. On Windows, os.path.join is an alias for ntpath.join, which always uses \. When dealing with URLs, we always want forward slashes, regardless of platform, so posixpath.join should be used to build URL paths.